Prompt the user before wiping their handlers.json file

bryceitoc9 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Infrastructure (change which improves the lifecycle management (CI/CD, build, package, deploy, lint, etc) of the application, but does not change functionality.)
- [ ] Technical debt (change which improves the maintainability of the codebase, but does not change functionality)
- [ ] Testing (change which modifies or adds test coverage, but does not change functionality)
- [ ] Documentation (change which modifies or adds documentation, but does not change functionality)

## Description

Previously, the handlers.json file would be overwritten if a user clicks the "Configure" CodeLens and either of the following happened:
* The JSON symbol provider hasn't loaded in the allotted time (1.5 seconds)
* The JSON file does not have a single valid top-level symbol

This increases the wait time to 2 seconds and will instead prompt a user if there's an issue. This will allow the user to edit and retry a load, to cancel the validation, or to overwrite.

This change also adds the `environmentVariables` snippet to the file if it isn't present and a handler is.

## Motivation and Context
This resolves the edge case of a user having their JSON file wiped if invalid or if the computer can't load the symbol provider in time

## Related Issue(s)

https://github.com/aws/aws-toolkit-vscode/issues/224

## Testing
* Tested prompt by changing the number of retries on the `loadSymbols` function to 0 and confirming the three options worked as expected.
* Tested JSON file adder by removing various keys from the handlers.json file, clicking "Configure", and confirming the additions of missing fields.

## Screenshots (if appropriate)

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

awschristou commented on src/lambda/local/configureLocalLambda.ts:
```diff
@@ -131,31 +132,66 @@ function getTabSize(editor?: vscode.TextEditor): number {
+    if (maxAttempts <= 0) {
+        // JSON symbol provider hasn't loaded and/or file exists but is unparseable (default 2 seconds total)
+        // (e.g. file is blank, only contains `{}`, or does not have any valid top level keys)
+        const localize = nls.loadMessageBundle()
```
Move this line up just below the import statements. That's how we set up `localize` in other files

awschristou commented on src/lambda/local/configureLocalLambda.ts:
```diff
@@ -131,31 +132,66 @@ function getTabSize(editor?: vscode.TextEditor): number {
         return symbols
     }
 
+    if (maxAttempts <= 0) {
```
This method's responsibility went from "get symbols from a file" to "get symbols from a file but if it can't then ask the user what they want to do", meaning an additional concern was added. This also makes the return value less clear ("array OR boolean with specific meaning). 

What if we leave the return value as it was ("array or undefined") and make this prompting logic the responsibility of the caller? The function remains a "symbol retriever" and our caller can decide whether or not it wants to call into the method again, and the caller can also decide if it will clobber the file.

bryceitoc9 commented:
Implemented @awschristou 's suggestions and added an interface for handling context as per @mpiroc 's suggestion. Still working on tests to take advantage of the context interface.

mpiroc commented on src/lambda/local/configureLocalLambda.ts:
```diff
@@ -29,23 +32,42 @@ export interface HandlersConfig {
 }
 
+export interface ConfigureLocalLambdaContext {
+    showTextDocument(document: vscode.TextDocument, column?: vscode.ViewColumn | undefined,
```
This works, but there's a shortcut:

```typescript
export interface ConfigureLocalLambdaContext {
    showTextDocument: typeof vscode.window.showTextDocument
    executeCommand: typeof vscode.commands.executeCommand
    showInformationMessage: typeof vscode.window.showInformationMessage
}
```

mpiroc commented on src/lambda/local/configureLocalLambda.ts:
```diff
@@ -131,31 +153,77 @@ function getTabSize(editor?: vscode.TextEditor): number {
 
-async function prepareConfig(editor: vscode.TextEditor, handler: string): Promise<boolean> {
-    const symbols: vscode.DocumentSymbol[] | undefined = await loadSymbols(editor.document.uri)
+async function prepareConfig(editor: vscode.TextEditor, handler: string,
```
Nit: Throughout this changeset, when splitting arguments accross multiple lines, use this style:

```typescript
async function prepareConfig(
    editor: vscode.TextEditor,
    handler: string,
    context: ConfigureLocalLambdaContext
)
```

mpiroc commented on src/lambda/local/configureLocalLambda.ts:
```diff
@@ -131,31 +153,77 @@ function getTabSize(editor?: vscode.TextEditor): number {
+    let shouldOverwrite: boolean = false
+    let shouldLoop: boolean = !symbols
+
+    while (!symbols && shouldLoop) {
```
Minor: This function is getting pretty long. Suggest to break out some of the functionality into helper methods as you see fit.
